### ISSUE TITLE
Can HTTP2 Server Push be implemented yet?

### ISSUE THREAD

**Issue by chrisblakley** [chrisblakley/Nebula#1343]
I started reading about it but I can't find any actual implementation examples...

http://caniuse.com/#feat=http2

Is it a `<link>` `rel` tag besides just `preload` or `prefetch`? Check in the Network tab for `push` to see if it's working.

![networktimeline](https://cloud.githubusercontent.com/assets/6349678/26509438/c6ef81c6-4226-11e7-908c-8f8f6f11bf3a.png)

Not interested in using JavaScript for this. I'll either recommend a directive approach or use standard HTML tags/attributes.

**Comment by chrisblakley** [comment]
I enhanced `prebrowsing()` to include `preload` on the `prefetch` links, but there could be more to it. Also, this does not push images.

**Comment by chrisblakley** [comment]
Simply using `rel="preload"` directly on the resource triggers Chrome violations:

```
The resource https://gearside.com/nebula/wp-content/themes/Nebula-master/assets/img/meta/favicon-16x16.png was preloaded using link preload but not used within a few seconds from the window's load event. Please make sure it wasn't preloaded for nothing.
```

My guess is that it is being preloaded at the same time as it is used, like this:
```html
<link rel="shortcut icon preload prefetch" type="image/png" sizes="16x16" href="https://gearside.com/nebula/wp-content/themes/Nebula-master/assets/img/meta/favicon-16x16.png" />
```

I'm not exactly sure how to trigger the HTTP2 Push still...

**Comment by chrisblakley** [comment]
Phew! 3 hours for about 7 lines of code.

The meat and potatoes of this is that you need to send this header for each local style/script that you want to push:
```php
header('Link: </nebula/wp-content/themes/Nebula-master/style.css>; rel=preload; as=style, false);
```

The infuriating part was trying to figure out how to send the header **after** styles and scripts were enqueued. The thing that **none of the blog posts tell you** is that you need to enable output buffering to do this!

So, on `init`, simply trigger this: `ob_start();`

So now, Nebula pushes all local registered/enqueued styles and scripts. It's up to both the browser and server to support it, but more and more will eventually.

Next up: include a few images with the server push... like favicons.

**Comment by chrisblakley** [comment]
As a note:

HTTP2 Server Push does not put any files into the cache, so it is recommended to use Service Workers to manage the cache. I'm not sure how to implement a Service Worker into WordPress/Nebula yet...

I may even disable HTTP2 Server Push until that is figured out...

**Comment by chrisblakley** [comment]
I just did more testing and after 25 runs with webpagetest.org with variations of HTTP2 Server Push on and off, I found (according to the numbers) that it was faster **without**... Again, it was so minuscule that it is tough to say definitively.

Average of four runs sampled from each:
 - Enabled: http://gifted-magpie.w4.poopy.life/
  - Initial: 2.53125s
  - Repeat: 1.51225s
 - Disabled: http://healthy-reindeer.w4.poopy.life/
  - Initial: 2.293s
  - Repeat: 1.48625s

This is impossible to test on my cloud hosting because the server response time varies wildly (similar to a shared host). So I tested these on poopy.life servers.
